File: benchmarks/59-polynomial-roots/polynomial_roots.py
#!/usr/bin/env python3
"""Experiment 59 -- High-degree polynomial roots (numpy column).

Same kernels as ``polynomial_roots.m``, same order and labels.

Baseline is compiled (numpy.roots -> companion-matrix eigenvalues via LAPACK).
Both sides compute the same roots, so the order-invariant check Total[|root|]
agrees to 4 places even at degree 100.  Dense coefficients follow denseUPoly
from data.m (built here in pure Python so we do not pay sympy's import cost to
reach numpy.roots); NRoots' count is len(roots).
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import numpy as np

from harness import bench, check, require

require(["numpy:roots"])


def dense_coeffs_desc(n):
    """Descending coefficient list of denseUPoly[n]: ((k^2+3k+1) mod 17) x^k + 1."""
    c = [((k * k + 3 * k + 1) % 17) for k in range(n + 1)]
    c[0] += 1
    return c[::-1]


def r(x, p):
    return int(np.floor(float(x) * 10 ** p + 0.5))


c20 = dense_coeffs_desc(20)
c50 = dense_coeffs_desc(50)
c100 = dense_coeffs_desc(100)
c40 = dense_coeffs_desc(40)
cRootUnity = [1] + [0] * 63 + [-1]                 # x^64 - 1
cWilk = np.poly(np.arange(1, 16))                  # (x-1)...(x-15)

bench("NSolve dense degree 20", lambda: np.roots(c20))
check("NSolve dense degree 20", r(np.abs(np.roots(c20)).sum(), 4))

bench("NSolve dense degree 50", lambda: np.roots(c50))
check("NSolve dense degree 50", r(np.abs(np.roots(c50)).sum(), 4))

bench("NSolve dense degree 100", lambda: np.roots(c100))
check("NSolve dense degree 100", r(np.abs(np.roots(c100)).sum(), 4))

bench("NSolve roots of unity degree 64", lambda: np.roots(cRootUnity))
check("NSolve roots of unity degree 64", r(np.abs(np.roots(cRootUnity)).sum(), 4))

bench("NSolve Wilkinson degree 15", lambda: np.roots(cWilk))
check("NSolve Wilkinson degree 15", r(np.abs(np.roots(cWilk)).sum(), 3))

bench("NRoots dense degree 40", lambda: np.roots(c40))
check("NRoots dense degree 40", int(len(np.roots(c40))))

File: benchmarks/59-polynomial-roots/polynomial_roots.m
(* ==========================================================================
   Experiment 59 -- High-degree polynomial roots (group E)
   ==========================================================================
   WHAT IT MEASURES.  NSolve / NRoots on dense univariate polynomials --
   src/numerical_roots/nroots.c, nroots_aberth.c (Aberth-Ehrlich), nroots_jt.c
   (Jenkins-Traub).  Experiment 14 finds a single scalar root; this finds ALL
   roots of degree 20 -> 100, the many-root regime where the companion-matrix /
   simultaneous-iteration methods live.

   Baseline is numpy.roots (companion-matrix eigenvalues via LAPACK), so both
   sides compute the same roots and the order-invariant check Total[|root|]
   agrees to 4 places even at degree 100.  The dense coefficients come from
   denseUPoly (data.m), shared with the Python side; roots-of-unity and the
   Wilkinson polynomial add a unit-circle case and an ill-conditioned case.

   Checks: sum of |root| (order-invariant, exercises the root-finder, unlike the
   Vieta-trivial sum of roots), or the root count.
   ========================================================================== *)

Get["../harness.m"];
Get["../data.m"];

require[{"NSolve", "NRoots"}];

p20 = denseUPoly[20, x];
p50 = denseUPoly[50, x];
p100 = denseUPoly[100, x];
p40 = denseUPoly[40, x];
pRootUnity = x^64 - 1;
pWilk = Product[(x - i), {i, 1, 15}];

bench["NSolve dense degree 20", NSolve[p20 == 0, x];];
check["NSolve dense degree 20", Round[10^4 Total[Abs[x /. NSolve[p20 == 0, x]]]]];

bench["NSolve dense degree 50", NSolve[p50 == 0, x];];
check["NSolve dense degree 50", Round[10^4 Total[Abs[x /. NSolve[p50 == 0, x]]]]];

bench["NSolve dense degree 100", NSolve[p100 == 0, x];];
check["NSolve dense degree 100", Round[10^4 Total[Abs[x /. NSolve[p100 == 0, x]]]]];

bench["NSolve roots of unity degree 64", NSolve[pRootUnity == 0, x];];
check["NSolve roots of unity degree 64",
      Round[10^4 Total[Abs[x /. NSolve[pRootUnity == 0, x]]]]];

bench["NSolve Wilkinson degree 15", NSolve[pWilk == 0, x];];
check["NSolve Wilkinson degree 15", Round[10^3 Total[Abs[x /. NSolve[pWilk == 0, x]]]]];

bench["NRoots dense degree 40", NRoots[p40 == 0, x];];
check["NRoots dense degree 40", Length[NRoots[p40 == 0, x]]];
